### PR TITLE
Refactor driver

### PR DESCRIPTION
**Pull request by calebmer**
Some initial thoughts. The driver is still very shallow though. Only works on a collection level…

### CODE REVIEW

**Comment by calebmer** [comment]
Not ready to be merged.

**Comment by svmnotn** [comment]
if you could add the rustfmt, and .gitignore changes from #3 into this, since it will probably get merged first, that would be nice.

Also don't forget `&self`

**Comment by calebmer** [comment]
Just commit those to the `develop` branch directly as a chore type.

**Comment by svmnotn** [comment]
done, updating #3 to merge nicely with develop was a chore.... I never want to do that again...

File: src/error.rs [hunk 1]
```diff
@@ -4,14 +4,24 @@
...
   /// A message providing more detail beyond the error code.
-  message: String,
+  message: &'a str,
   /// A hint to the user on what to do next to try and avoid the error
   /// happening again. This is optional.
-  hint: Option<String>
+  hint: Option<&'a str>
+}
+
+impl<'a> Error<'a> {
```
Comment (svmnotn) [inline src/error.rs:17]: impl Error for Error

Comment (calebmer) [inline src/error.rs:17]: I'll do that in a new branch. Eventually 😊

**Comment by calebmer** [comment]
Ready for review.

File: src/error.rs [hunk 2]
```diff
@@ -4,14 +4,24 @@
...
+  pub fn unimplemented(message: &'a str) -> Self {
```
Comment (svmnotn) [inline src/error.rs:29]: cant this fn just take the ErrorCode so that it is more generic?

**Comment by svmnotn** [comment]
Ok, other than the unimplemented fn in Error, it should be ok to merge

File: src/error.rs (hunk 2 above)
Comment (calebmer) [inline src/error.rs:29]: We can add another function, I just wanted a convenience function for writing unimplemented errors.

**Pull request merged**